lean/bitvec.lean

```diff
@@ -0,0 +1,375 @@
+  begin
+    apply bitvec_intro,
+    simp [append, split_at', cong_val, split_at],
+    apply (trunc_add_ushr H)
+  end
+
+end listlike
+
+section arith
+  -- Arithmetic operations on bitvectors
+  variable {n : ℕ}
+
+  -- 2s complement negation
+  protected def neg (x : bitvec n) : bitvec n :=
+    ⟨if x.val = 0 then 0 else 2^n - x.val,
+     begin
+       by_cases (x.val = 0),
+       { simp [h], apply pos_pow_of_pos, dec_trivial_tac },
+       { simp [h],
+         have pos : 0 < 2^n - x.val, { apply nat.sub_pos_of_lt x.property },
+         have x_val_pos: 0 < x.val, { apply nat.pos_of_ne_zero h },
+         apply sub_lt_of_pos_le x.val (2^n) x_val_pos,
+         apply le_of_lt x.property,
+       }
+      end⟩
+
+  -- Add with carry (no overflow)
+  def adc (x y : bitvec n) (c : bool) : bitvec n × bool :=
+    let c₁ := if c then 1 else 0,
+        r  := x.val + y.val + c₁ in
+    ⟨ ⟨r % 2^n, by simp ⟩, r ≥ 2^n ⟩
+
+  protected def add (x y : bitvec n) : bitvec n := (adc x y ff).1
+
+  -- Subtract with borrow
+  def sbb (x y : bitvec n) (b : bool) : bool × bitvec n :=
+    let b₁ : bitvec n := if b then 1 else 0,
+        r  := match bitvec.adc x (bitvec.neg y) ff with
+              | (z, b₂) := bitvec.adc z (bitvec.neg b₁) ff
+              end
+    in ⟨ if b then y.val + 1 > x.val else y.val > x.val , r.1 ⟩
+
+  -- Usual arithmetic subtraction
+  protected def sub (x y : bitvec n) : bitvec n := (sbb x y ff).2
+
+  instance : has_add (bitvec n)  := ⟨bitvec.add⟩
+  instance : has_sub (bitvec n)  := ⟨bitvec.sub⟩
+  instance : has_neg (bitvec n)  := ⟨bitvec.neg⟩
+
+  protected def mul (x y : bitvec n) : bitvec n :=
+    ⟨ (x.val * y.val) % 2^n, by simp ⟩
+
+  instance : has_mul (bitvec n)  := ⟨bitvec.mul⟩
+
+  def bitvec_pow {n:ℕ} : bitvec n → ℕ → bitvec n
+    | x k := ⟨ (x.val^k) % 2^n, by simp ⟩
+
+  instance bitvec_has_pow {n:ℕ} : has_pow (bitvec n) ℕ := ⟨bitvec_pow⟩
+
+end arith
+
+section comparison
+  -- Comparison operations, including signed and unsigned versions
+  variable {n : ℕ}
+
+  def uborrow (x y : bitvec n) : bool := prod.fst (sbb x y ff)
+
+  def ult (x y : bitvec n) : Prop := uborrow x y
+  def ugt (x y : bitvec n) : Prop := ult y x
+
+  def ule (x y : bitvec n) : Prop := ¬ (ult y x)
+  def uge (x y : bitvec n) : Prop := ule y x
+
+  def sborrow : Π {n : ℕ}, bitvec n → bitvec n → bool
+  | 0        _ _ := ff
+  | (succ n) x y :=
+    match (msb x, msb y) with
+    | (tt, ff) := tt
+    | (ff, tt) := ff
+    | (ff, ff) := uborrow x y
+    | (tt, tt) := uborrow (bitvec.neg y) (bitvec.neg x) -- -x < -y when y < x
+    end
+
+  def slt (x y : bitvec n) : Prop := sborrow x y
+  def sgt (x y : bitvec n) : Prop := slt y x
+  def sle (x y : bitvec n) : Prop := ¬ (slt y x)
+  def sge (x y : bitvec n) : Prop := sle y x
+
+end comparison
+
+section conversion
+  -- Operations for converting to/from bitvectors
+  variable {α : Type}
+
+  protected def of_nat (n : ℕ) (x:ℕ) : bitvec n := ⟨ x % 2^n, by simp ⟩
```

Comment:
If we move this up, we could just this earlier to fundings that call mod and simp.